### PR TITLE
Pretty printer preserves comments

### PR DESCRIPTION
**Pull request by nhuurre**
Another shot at #93 . This is less hacky than #603  and keeps _all_ comments (the comments with "unexpect" placement get pushed a few lines forward).

The strategy here is _not_ to attach the comments to AST nodes but to dump them all into a single list with location metadata for each. Since AST nodes already have location metadata (for reporting semantic errors) the pretty printer can use that to interweave the comments with the code it outputs.

This is draft because unfortunately something goes wrong with `#include` and I can't figure out why.

## Release notes

Correct canonicalizing/auto-formatting of commented Stan programs; comments no longer discarded.

## Copyright and Licensing
Copyright holder: Niko Huurre
By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by nhuurre** [comment]
Ok, I figured it out. The location metadata coming out the parser is a bit off, the `begin_loc` of a statement is actually at the end of the previous statement and that means some hacks are required to correctly format the whitespace in between.

Anyway, this is ready for review.
I don't understand what's causing the Jenkins failure but it seems unrelated.

**Comment by rok-cesnovar** [comment]
Sorry Niko, just saw that this has been waiting for 2 weeks+.

@rybern would you be able to have a look at this, given that you already discussed solutions/reviewed in #603? Else I can take a look.

**Comment by rybern** [comment]
Sorry, probably not going to be able to help with this in a timely manner @rok-cesnovar

File: test/integration/good/vec-expr/pretty.expected [hunk 1]
```diff
@@ -115,57 +116,67 @@ generated quantities {
...
+  
+  vector[100] e = [1.0, 2.0, 3.0, y, y, y, 7, 8, 9.0, 10, 1.0, 2.0, 3.0, 
+                   y, y, y, 7, 8, 9.0, 10, 1.0, 2.0, 3.0, y, y, y, 7, 
+                   8, 9.0, 10, 1.0, 2.0, 3.0, y, y, y, 7, 8, 9.0, 10, 
+                   1.0, 2.0, 3.0, y, y, y, 7, 8, 9.0, 10, 1.0, 2.0, 3.0, 
+                   y, y, y, 7, 8, 9.0, 10, 1.0, 2.0, 3.0, y, y, y, 7, 
+                   8, 9.0, 10, 1.0, 2.0, 3.0, y, y, y, 7, 8, 9.0, 10, 
+                   1.0, 2.0, 3.0, y, y, y, 7, 8, 9.0, 10, 1.0, 2.0, 3.0, 
+                   y, y, y, 7, 8, 9.0, 10]';
+
```
Comment (rok-cesnovar) [inline test/integration/good/vec-expr/pretty.expected:171]: Is this additional space intentional?

**Review by rok-cesnovar — COMMENTED** [review]
Looks good to me, I do have a question about all the new newlines (see pretty.expected changes). Are they intentional and could we avoid adding them?

File: test/integration/good/vec-expr/pretty.expected (hunk 1 above)
Comment (nhuurre) [reply test/integration/good/vec-expr/pretty.expected:171]: Yes. The empty line is also present in the original source file:
https://github.com/stan-dev/stanc3/blob/c620028502caadf14807be101341c1da54df6d1f/test/integration/good/vec-expr/vec_expr_big.stan#L66-L68
Most of these empty lines are from empty lines in the source but there's a couple of places where new unintentional lines appear. The problem is that when you open a local block with brace `{` the parser does not record the location of the brace but you'd usually put it on its own line like 
```stan
real x;
{
  real y = 2.0 + 2.0;
  x = y*y;
}
```
and now the pretty printer thinks there was an empty line...so it adds a superfluous line break before the block...

Comment (WardBrian) [reply test/integration/good/vec-expr/pretty.expected:171]: Is this problem recurrent, or is it only on the first pass? 

Put another way, I have a problem `prog1.stan` which uses local blocks. If I pretty-print it to produce `prog2.stan`, it will have an additional space per block, understandable. If I now pretty-print `prog2.stan` to produce `prog3.stan`, with `prog3` be identical to `prog2`, or will it have even more whitespace?

Comment (nhuurre) [reply test/integration/good/vec-expr/pretty.expected:171]: A sequence of multiple blank lines collapses to a single line. `prog3.stan` and `prog2.stan` should be identical.

**Comment by WardBrian** [comment]
@nhuurre - would you mind if I take a crack at adapting what you've done to line comments?

**Comment by nhuurre** [comment]
@WardBrian 
Sorry, I should have gotten back to this earlier. Feel free try if you want. Pushed a commit with partial progress right now.
I think you need to use `Format.pp_force_newline` to guarantee a line break after the comment without messing up the formatting but when I tried I got an empty line after every line comment.

File: src/frontend/Pretty_printing.ml [hunk 2]
```diff
@@ -75,18 +85,20 @@ let pp_comment ppf (lines, {Middle.Location_span.begin_loc= {col_num; _}; _}) =
   let trim_tail col_num lines =
     match lines with [] -> [] | hd :: tl -> hd :: trim (col_num - 2) tl
   in
-  Fmt.pf ppf "@[<v>/*%a*/@]" Fmt.(list string) (trim_tail col_num lines)
+  if is_block then
+    Fmt.pf ppf "@[<v>/*%a*/@]" Fmt.(list string) (trim_tail col_num lines)
+  else Fmt.pf ppf "@[<v>/*%a */@]" Fmt.(list string) lines
```
[the remark this answers is not in the document]
Comment (nhuurre) [reply src/frontend/Pretty_printing.ml:90]: Yes, this just reproduces a block comment at the moment.
A `//` comment needs a new line at the end.

[the remark this answers is not in the document]
Comment (WardBrian) [reply src/frontend/Pretty_printing.ml:90]: It seems like newlines are preserved in most cases, it is only a few edge cases where weird things happen, like in test/integration/good/ode_good.stan we have this section:

```stan
functions {
  real[] harm_osc_ode(real t,
                      real[] y,         // state
                      real[] theta,     // parameters
                      real[] x,         // data
                      int[] x_int) {    // integer data
```

which fails in the naive solution I suggested above.

[the remark this answers is not in the document]
Comment (nhuurre) [reply src/frontend/Pretty_printing.ml:90]: Exactly. That's where you need `Format.pp_force_newline`. But I'm afraid that knowing if you're in that context requires a large refactoring of the printer code.

[the remark this answers is not in the document]
Comment (WardBrian) [reply src/frontend/Pretty_printing.ml]: I've updated the tests which work for now, leaving out the exceptions so the tests will still fail in CI. This will require some thought, but I think we could try a special case if a comma and a line comment are on the same line?

[the remark this answers is not in the document]
Comment (nhuurre) [reply src/frontend/Pretty_printing.ml]: The difference is `pp_list_of` (inside parens) vs `pp_list_of_statements` (at toplevel). Those don't share as much code as I thought so maybe it's not so bad.

[the remark this answers is not in the document]
Comment (WardBrian) [reply src/frontend/Pretty_printing.ml]: Hopefully it is simpler then. Feel free to ping me if you want help/someone else to look at the problem

**Review by WardBrian — APPROVED** [review]
LGTM! This is a really great feature, thanks for all your work!
